torch/csrc/jit/passes/decompose_ops.cpp

```diff
@@ -21,8 +27,39 @@ static bool DecomposeOps(Block* block) {
     for (auto sub : it->blocks()) {
       DecomposeOps(sub);
     }
+    if (it->matches("aten::linear(Tensor input, Tensor weight, Tensor? bias) -> Tensor")) {
+      Value* input = it->namedInput(attr::input);
+      Value* weight = it->namedInput(attr::weight);
+      Value* bias = it->namedInput(attr::bias);
+      auto input_type = input->type()->cast<DimensionedTensorType>();
+      if (!input_type) {
+        // if the input type is not specialized, don't do decomposition
+        continue;
+      }
+
+      decomposed = true;
+      WithInsertPoint guard(*it);
 
-    if (it->matches(
+      Graph* graph = it->owningGraph();
+      std::shared_ptr<Graph> d_graph;
+      int ndim = input_type->dim();
+      Value* new_output = nullptr;
+      if (ndim == 2 && bias->type()->isSubtypeOf(TensorType::get())) {
+        // if ndim == 2 and bias is statically defined, dispatch to addmm decomposition
```

Comment:
_statically defined_ is not very well defined here. You're only checking that we can refine its type, which doesn't tell you much.

Reply:
This is where I steal it from the [fuser code](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/graph_fuser.cpp#L156), where is tells you if it is statically defined (statically undefined needs more check though). because we are only check the input value types, I guess it would be sufficient to distinguish it. 

The next step after this PR, I will move the `batchnorm` and `layernorm` decomposition to here as well, then I can use the util function defined above.

Reply:
My point is that the term _statically defined_ is not used correctly. I'm not even 100% sure what does that mean

Reply:
From my point of view, `statically defined` tensor argument means that the value passed in is neither a `None` constant nor a `Optional[Tensor]` type, which we could statically know the type passed in rather than a dynamically unknown optional type. So `isSubtypeOf(TensorType)` should guarantee this definition